[Mobile] Update the bottom sheet header (#34309)
WordPress/gutenberg: packages/components/src/mobile/bottom-sheet/header/README.md

## Usage

```jsx
import { BottomSheet } from '@wordpress/components';

export default = () => (
	<BottomSheet>
		<BottomSheet.Header>
			<BottomSheet.Header.BackButton onPress={ () => {} } />
			<BottomSheet.Header.Title>A Sheet Title</BottomSheet.Header.Title>
			<BottomSheet.Header.ApplyButton onPress={ () => {} } />
		</BottomSheet.Header>
	</BottomSheet>
);
```

## BottomSheet.Header

Provides structural styles for left-center-right layout for header UI.

## BottomSheet.Header.Title

Displays a styled title for a bottom sheet.

## BottomSheet.Header.ApplyButton

Displays a styled button to apply settings of bottom sheet controls.

### Props

#### onPress

Callback invoked once the button is pressed.

## BottomSheet.Header.BackButton

Displays a styled button to navigate backwards from a bottom sheet.

### Props

#### onPress

Callback invoked once the button is pressed.

## BottomSheet.Header.CancelButton

Displays a styled button to dismiss a full screen bottom sheet.

### Props

#### onPress

Callback invoked once the button is pressed.

## BottomSheet.Header.CloseButton

Displays a styled button to dismiss a full screen bottom sheet.

### Props

#### onPress

> Callback invoked once the button is pressed.

Reviewer: Though the description here is the same for both of these components, I'm wondering about their intended semantics, and whether they really are the same. I think that Cancel implies "undoing" some changes that were made, while Close does not imply that. Otoh, what actually happens is entirely dependent on the implementation in the callback. Maybe we can find a way to define the difference between these components in the readme to guide developers that use these components toward a more consistent usage?

Reviewer 2: > Otoh, what actually happens is entirely dependent on the implementation in the callback. 

That was my original thought with the documentation for the `onPress` callback. The current description is very much focused on what _does_ occur, not the intended usage.

> Maybe we can find a way to define the difference between these components in the readme to guide developers that use these components toward a more consistent usage?

Counter to the current documentation (I modified) and my aforementioned original thought, I do understand and agree there is value in describing the intent. The description of the components themselves differ somewhat from the `onPress` prop description. E.g. `ApplyButton` states "Displays a styled button to apply settings of bottom sheet controls," which describe the intent to apply the settings. 

If we want to update the documentation to better describe the intent of the components and props, that is fine with me. Apologies if my previous changes removed too much of that, and thus creating more work to add the intent back. 😄

Reviewer: > Apologies if my previous changes removed too much of that, and thus creating more work to add the intent back.

No worries! I think it's great to discuss this and I'm happy this is being addressed. The usage of these patterns will certainly grow, so it's valuable to consider many paths. There is a tendency (and maybe a goal?) to make changes to block settings immediately reflected in the block rendering (i.e. no "confirm" or "cancel" needed). I think this can't cover the case for any fullscreen sheets, though, so this may be _at least one_ source of tension. Historically, though, this may have been addressed in a somewhat ad-hoc way, giving rise to some redundancy in naming — but that's just my guess.

In any case, removing truly synonymous components and clarifying the usage of the remaining ones is a great idea. Also, I especially like [the idea mentioned here](https://github.com/WordPress/gutenberg/pull/34309#discussion_r697477701) to add `iosText` as a prop. This accurate signalling provides immediate clarity, whereas the current form feels like it is violating the principle of least surprise, imo.

> # Header

Reviewer 2: nitpick: My personal opinion is the current documentation is somewhat verbose and unfocused. I put together [an alternative example](https://github.com/WordPress/gutenberg/pull/34339) for consideration, but I recognize it is a lot of change and quite opinionated. Please feel free to ignore it entirely or take anything from it you deem valuable. 🙇🏻

Author: Thanks for the suggestions in the documentation.

> Header components is meant to be used to compose the header inside the BottomSheet.

Reviewer 3: ```suggestion
`BottomSheet.Header` should be used to add a header to the top of a `BottomSheet` component. It makes several other components available, which can then be used to compose the header's content.
```
It took me a bit to fully process how this component works, so think adding a bit more detail to the intro might help with clarity. WDYT?

### Usage

> See 

Reviewer 3: ```suggestion
```

It looks like this can either be removed or perhaps there was a missing referencing?

Author: Thanks for the suggestion @SiobhyB The documentation section got converted updated in https://github.com/WordPress/gutenberg/pull/34339.

```jsx
/**
 * External dependencies
 */
import { View, Text } from 'react-native';
import { useNavigation } from '@react-navigation/native';

/**
 * WordPress dependencies
 */
import { BottomSheet } from '@wordpress/components';

const ExampleControl = () => {

    const goBack = () => {};
    const applySettings = () => {};

	return (
		<BottomSheet.SubSheet>
			<>
				<BottomSheet.Header>
					<BottomSheet.Header.BackButton onPress={ goBack } />
					<BottomSheet.Header.Title>{ 'Howdy' }</BottomSheet.Header.Title>
                    <BottomSheet.Header.ApplyButton onPress={ applySetting } />
				</BottomSheet.Header>
				<View paddingHorizontal={ 16 }>
					<Text>{ 'World' }</Text>
				</View>
			</>
		</BottomSheet.SubSheet>
	);
};

export default ExampleControl;
```

### Props
Header component doesn't have any props. 

### Other Components
Other components that Header component makes available. 

#### ApplyButton 
```
<Header.ApplyButton onPress={ goBack } />
```

The apply button is used to apply settings of the bottom sheet control.

##### onPress 

> use to pass a call back once the ApplyButton is clicked. 

Reviewer 3: ```suggestion
Use to pass a call back once the ApplyButton is clicked. 
```

#### BackButton 
```
<Header.BackButton onPress={ goBack } />
```

The back button is used to send the user back to the previous sheet. 

##### onPress 

> use to pass a call back once the BackButton is clicked. 

Reviewer 3: ```suggestion
Use to pass a call back once the BackButton is clicked. 
```

#### CancelButton 
```
<Header.CancelButton onPress={ goBack } />
```

The cancel button is used to send the user back to the previous sheet. Closes the sheet. Use this if you are using a full screen bottom sheet. 

##### onPress 

> use to pass a call back once the CancelButton is clicked. 

Reviewer 3: ```suggestion
Use to pass a call back once the CancelButton is clicked. 
```
